[ML] Add auto generated drill down link to Discover for Anomaly explorer table

qn895 opened the pull request:
## Summary

This PR adds auto generated link to analyze records in Discover to the Anomaly table. Previously, the `Raw data` link will only show up if the job was created using the predefined templates.

https://user-images.githubusercontent.com/43350163/144934301-37559313-5243-45d8-9ab2-443fb35c5b6d.mov

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

darnautov commented on x-pack/plugins/ml/public/application/components/anomalies_table/links_menu.js:
```diff
@@ -53,6 +53,51 @@ class LinksMenuUI extends Component {
+    const dataViewId = (await getDataViewIdFromName(index)) || index;
+
+    const record = this.props.anomaly.source;
+    const from = timeFormatter(record.timestamp); // e.g. 2016-02-08T16:00:00.000Z
```
Could you please use [this hook](https://github.com/elastic/kibana/blob/e10ac3c56e6c0e5144dfbbd8fe8b9d56bc1e643d/x-pack/plugins/ml/public/application/contexts/kibana/use_field_formatter.ts#L21) instead? It utilizes the `fieldFormats` service.

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/120450/commits/e982410ce8d3231a42b0a3f40821be9a6b317043

darnautov replied:
Great 👍 Please bear in mind that default date formatting in Kibana is not `2016-02-08T16:00:00.000Z`. You need to pass the required options as a second argument for `useFieldFormatter `

qn895 replied:
Looks like there's a timezone issue with the fieldFormatter services. I reverted back to use the previous timeFormatter for now https://github.com/elastic/kibana/pull/120450/commits/d03dfd4aa2ff76d933a7ee09c3fd416dd3245dde while investigating this further. As the from/to here is not shown to the users (it's only for the url), I think it should be fine.

darnautov commented on x-pack/plugins/ml/public/application/components/anomalies_table/links_menu.js:
```diff
@@ -53,6 +53,51 @@ class LinksMenuUI extends Component {
+    const from = timeFormatter(record.timestamp); // e.g. 2016-02-08T16:00:00.000Z
+    const to = timeFormatter(record.timestamp + record.bucket_span * 1000);
+
+    let luceneQuery = '';
```
why Lucene but not KQL? 🤔

qn895 replied:
Good point! I updated it to KQL now. Thanks.

darnautov commented on x-pack/plugins/ml/public/application/components/anomalies_table/links_menu.js:
```diff
@@ -53,6 +53,51 @@ class LinksMenuUI extends Component {
+      sort: [['timestamp, asc']],
+    });
+
+    window.open(url);
```
I think we should avoid using `window.open`. Shall we open in the same tab using `navigateToApp`?

peteharverson replied:
Is it possible to have the context menu item click open in a new tab using e.g. keyboard shortcuts?

qn895 replied:
Updated to a href here https://github.com/elastic/kibana/pull/120450/commits/d6e4b9adae89b65416ff57ac59a43136b49e508b

darnautov commented on x-pack/plugins/ml/public/application/components/anomalies_table/links_menu.js:
```diff
@@ -411,6 +456,22 @@ class LinksMenuUI extends Component {
+          }}
+          data-test-subj={`mlAnomaliesListRowAction_rawDataButton`}
+        >
+          {i18n.translate('xpack.ml.anomaliesTable.linksMenu.rawData', {
```
For react components, `FormattedMessage` is preferred according to the Kibana guidelines

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/120450/commits/d6e4b9adae89b65416ff57ac59a43136b49e508b

peteharverson commented on x-pack/plugins/ml/public/application/components/anomalies_table/links_menu.js:
```diff
@@ -53,6 +53,51 @@ class LinksMenuUI extends Component {
+
+    const record = this.props.anomaly.source;
+    const from = timeFormatter(record.timestamp); // e.g. 2016-02-08T16:00:00.000Z
+    const to = timeFormatter(record.timestamp + record.bucket_span * 1000);
```
I think it would be good to use the same logic defined in `openCustomUrl` in this file as we have for user-defined custom URLs with 'auto' time range setting for determining the `from` and `to` for the Discover link. This uses the aggregation interval set in the table:

- day - show full day in Discover
- hour - shows an hour either side of the hour in which the anomaly is in
- show all - shows just the selected bucket

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/120450/commits/d6e4b9adae89b65416ff57ac59a43136b49e508b

peteharverson replied:
There seems to be an issue when the Kibana timezone `dateFormat:tz` advanced setting is set to something other than the default (browser), where the time range opened in Discover doesn't correctly adjust for the timezone in the same way that the existing custom URL functionality does. I think the adjustment is added twice e.g. viewing by hour in table, timezone set to +8, anomaly at Feb 10 00:00, Discover opens to show Feb 10 07:00-09:59, where it should be Feb 9 23:00 - Feb 10 01:59.

qn895 replied:
I was able to reproduce this as well. This hopefully should be fixed with https://github.com/elastic/kibana/pull/120450/commits/d03dfd4aa2ff76d933a7ee09c3fd416dd3245dde.

peteharverson commented on x-pack/plugins/ml/public/application/components/anomalies_table/links_menu.js:
```diff
@@ -53,6 +53,51 @@ class LinksMenuUI extends Component {
+    const job = mlJobService.getJob(this.props.anomaly.jobId);
+
+    const index = job.datafeed_config.indices[0];
+    const dataViewId = (await getDataViewIdFromName(index)) || index;
```
I think you might have uncovered a bug in `getDataViewIdFromName`. 

Create two data views:
- cloudwatch-2019.*
- cloudwatch-2019.11.04

The create a job off the wildcard index pattern `cloudwatch-2019.*`. The new Discover action for my setup then opens Discover in the wrong data view `cloudwatch-2019.11.04`.

jgowdyelastic replied:
Yes, this is a bug.
The new dataviews service isn't as accurate as the old one. In other places I've fixed this by performing a second find on the results. e.g.
https://github.com/elastic/kibana/blob/c6d41d07d727c1257f482880f22564bae30449fb/x-pack/plugins/ml/server/models/data_recognizer/data_recognizer.ts#L628

I'll create a PR for this as there is another place this needs to be fixed.
https://github.com/elastic/kibana/pull/118006#discussion_r756187254

jgowdyelastic replied:
Raised https://github.com/elastic/kibana/pull/120737 to fix this

qn895 replied:
Thanks @jgowdyelastic for the quick turn around 🙏  I merged the main branch your latest changes and the fix works great.

peteharverson replied:
Confirmed that the target data view is set correctly with the changes merged in from https://github.com/elastic/kibana/pull/120737.

peteharverson commented:
This will be really useful functionality, making it easier to explore data around the time of anomalies without needing to set up user-defined custom URLs.

As well as re-using existing logic for calculating the drilldown time range, you should also look to re-use current custom URL functionality for edge cases such as:

- Data sets where the field names and values contain special characters - use `special_characters_2019` for testing. The code here isn't escaping special characters correctly.
- Categorization jobs - the code here passes on queries such as `mlcategory:15 AND machine.os.keyword:win 8`. There is probably no need to add it for categorization detectors, as the 'View examples' link opens up Discover anyway.

Can you also check it works with the `lat_long` geographic function - I think it should be ok as you only use the influencer field name and value in the query passed to Discover.

qn895 commented on x-pack/plugins/ml/public/application/components/anomalies_table/links_menu.js:
```diff
@@ -411,6 +456,22 @@ class LinksMenuUI extends Component {
+      items.push(
+        <EuiContextMenuItem
+          key={`auto_raw_data_url`}
+          icon="popout"
```
Updated here https://github.com/elastic/kibana/pull/120450/commits/d6e4b9adae89b65416ff57ac59a43136b49e508b

qn895 commented:
> * Data sets where the field names and values contain special characters - use `special_characters_2019` for testing. The code here isn't escaping special characters correctly.

Thanks for the hint. I have updated it to handle cases with special characters now.

https://user-images.githubusercontent.com/43350163/145253912-8b7ef3e1-7bea-4cdd-a792-89d9ff7bcfde.mov


> * Categorization jobs - the code here passes on queries such as `mlcategory:15 AND machine.os.keyword:win 8`. There is probably no need to add it for categorization detectors, as the 'View examples' link opens up Discover anyway.

Likewise, the categorization detectors will not be included in the query.

> Can you also check it works with the `lat_long` geographic function - I think it should be ok as you only use the influencer field name and value in the query passed to Discover.

You're right, it does work as expected.

https://user-images.githubusercontent.com/43350163/145253946-9d3bc59c-6aa7-426b-9376-c23eebbb485c.mov


I have also updated the `View in Discover` option to always show, not just when there's no custom url defined.

qn895 commented on x-pack/plugins/ml/public/application/components/anomalies_table/links_menu.tsx:
```diff
@@ -7,64 +7,158 @@
+          // No need to add categorization detectors to query
+          .filter((i) => i.influencer_field_name !== 'mlcategory')
+          .map(
+            (i) =>
```
Updated here https://github.com/elastic/kibana/pull/120450/commits/ee50d9c007f988e5726a80635f10efd75cdfdcb9

peteharverson commented:
> I have updated it to handle cases with special characters now.

The 'View in Discover' functionality is working with the special characters data now. But I noticed the existing custom URL functionality isn't working when the partition field name is `colon:name` - is there a way you could apply the code which works for the auto-generated link to the custom URL drilldown?

peteharverson commented on x-pack/plugins/ml/public/application/components/anomalies_table/links_menu.tsx:
```diff
@@ -7,64 +7,162 @@
+      if (record.influencers) {
+        kqlQuery = record.influencers
+          // No need to add categorization detectors to query
+          .filter((influencer) => influencer.influencer_field_name !== 'mlcategory')
```
I think if we are showing 'View examples' in the menu, then there is no point adding in 'View in Discover', as the 'View examples' opens up Discover anyway, with the query set to show docs for the anomalous category. I don't think 'View in Discover' showing the unfiltered docs adds any value to the existing 'View examples'. You should just check for `anomaly.entityName === 'mlcategory'` and not add View in Discover in this case.

![image](https://user-images.githubusercontent.com/7405507/145835463-d6462eb2-d4d4-4a03-a206-33977c183d34.png)

peteharverson replied:
Might be worth adding some sort of `isCategorizationAnomaly` function to `ml/common/util/anomaly_utils.ts`

peteharverson commented on x-pack/plugins/ml/common/util/anomaly_utils.ts:
```diff
@@ -42,6 +43,16 @@ export interface EntityField {
   operation?: EntityFieldOperation;
 }
 
+export interface Anomaly {
```
Could you adapt the `AnomaliesTableRecord` interface from `ml/common/types/anomalies.ts` rather than exporting a new `Anomaly` interface from here? The new `isCategorizationAnomaly` function is only called from the anomalies tables, so might be possible?

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/120450/commits/7d2a1f0901d0447ef620d8a8ef018ad65c08c5cc

peteharverson reviewed (approved):
Tested latest changes and overall LGTM. Just left a comment about reusing an existing interface. Happy to leave out of this PR the issue with the existing custom URLs when the field name contains a colon.

darnautov commented on x-pack/plugins/ml/common/types/anomalies.ts:
```diff
@@ -6,6 +6,7 @@
  */
 
 import { PARTITION_FIELDS, ANOMALY_RESULT_TYPE } from '../constants/anomalies';
+import { KibanaUrlConfig } from './custom_urls';
```
nit: 
```suggestion
import type { KibanaUrlConfig } from './custom_urls';
```

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/120450/commits/877fb6b603705d8b31271728f1e4935a4d1a7835

darnautov commented on x-pack/plugins/ml/common/util/anomaly_utils.ts:
```diff
@@ -14,7 +14,8 @@ import { i18n } from '@kbn/i18n';
 import { ANOMALY_SEVERITY, ANOMALY_THRESHOLD, SEVERITY_COLORS } from '../constants/anomalies';
-import type { AnomalyRecordDoc } from '../types/anomalies';
+import type { AnomaliesTableRecord, AnomalyRecordDoc } from '../types/anomalies';
+import { KibanaUrlConfig } from '../types/custom_urls';
```
```suggestion
import type { KibanaUrlConfig } from '../types/custom_urls';
```

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/120450/commits/877fb6b603705d8b31271728f1e4935a4d1a7835

darnautov commented on x-pack/plugins/ml/public/application/components/anomalies_table/links_menu.tsx:
```diff
@@ -7,64 +7,152 @@
+      const from = timeFormatter(earliestMoment.unix() * 1000); // e.g. 2016-02-08T16:00:00.000Z
+      const to = timeFormatter(latestMoment.unix() * 1000);
+
+      let kqlQuery = '';
+
+      if (record.influencers) {
+        kqlQuery = record.influencers
+          .map(
+            (influencer) =>
+              `${escapeForElasticsearchQuery(influencer.influencer_field_name)}:"${
+                influencer.influencer_field_values[0] ?? ''
+              }"`
+          )
+          .join(' AND ');
+      }
```
`@kbn/es-query”` package has a method to generate query string. did you try it? 
I think it's the preferred way over generating a query string manually.

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/120450/commits/877fb6b603705d8b31271728f1e4935a4d1a7835

darnautov commented on x-pack/plugins/ml/public/application/components/anomalies_table/links_menu.tsx:
```diff
@@ -7,64 +7,152 @@
+          language: 'kuery',
+          query: kqlQuery,
+        },
+        // interval,
```
do we need this comment?

qn895 replied:
Removed here https://github.com/elastic/kibana/pull/120450/commits/877fb6b603705d8b31271728f1e4935a4d1a7835

darnautov reviewed (approved):
LGTM, left some suggestions
